paddle/phi/kernels/gpu/multi_tensor_adam_kernel.cu

```diff
@@ -0,0 +1,278 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "paddle/phi/kernels/multi_tensor_adam_kernel.h"
+#include <assert.h>
+#include <cstdlib>
+#include <vector>
+#include "paddle/fluid/framework/tensor_util.h"
+#include "paddle/phi/backends/gpu/gpu_context.h"
+#include "paddle/phi/common/amp_type_traits.h"
+#include "paddle/phi/common/float16.h"
+#include "paddle/phi/common/scalar.h"
+#include "paddle/phi/core/dense_tensor.h"
+#include "paddle/phi/core/kernel_registry.h"
+#include "paddle/phi/core/tensor_utils.h"
+#include "paddle/phi/kernels/funcs/adam_functors.h"
+#include "paddle/phi/kernels/funcs/for_range.h"
+#include "paddle/phi/kernels/funcs/selected_rows_functor.h"
+#include "paddle/phi/kernels/multi_tensor_adam_utility_kernel.cuh"
+
+namespace phi {
+
+#define BLOCK_SIZE 512
+#define PARALLEL_SIZE 4
+
+template <typename T, typename MT, int N>
+struct CudaKernelFunctor {
+  __device__ __forceinline__ void operator()(int compute_group_size,
+                                             TensorAndBlockInf<N> tabi,
+                                             MT beta1,
+                                             MT beta2,
+                                             const MT* beta1_pow_,
+                                             const MT* beta2_pow_,
+                                             MT epsilon,
+                                             const MT* learning_rate,
+                                             bool mode,
+                                             bool multi_precision,
+                                             MT decay) {
+    MT lr = *learning_rate;
+    MT beta1_pow = *beta1_pow_;
+    MT beta2_pow = *beta2_pow_;
+
+    int tensor_id = tabi.tenosr_for_this_block[blockIdx.x];
+
+    int compute_group_idx = tabi.compute_group_for_this_block[blockIdx.x] +
+                            tabi.start_compute_group_this_tensor;
+
+    int n = tabi.sizes[tensor_id];
+    const T* g = static_cast<const T*>(tabi.grad[tensor_id]);
+    g += compute_group_idx * compute_group_size;
+    MT* mp;
+    T* p;
+    p = static_cast<T*>(tabi.tensors_addr[0][tensor_id]);
+    p += compute_group_idx * compute_group_size;
+    MT* m = static_cast<MT*>(tabi.tensors_addr[1][tensor_id]);
+    m += compute_group_idx * compute_group_size;
+    MT* v = static_cast<MT*>(tabi.tensors_addr[2][tensor_id]);
+    v += compute_group_idx * compute_group_size;
+
+    if (multi_precision) {
+      mp = static_cast<MT*>(tabi.tensors_addr[3][tensor_id]);
+      mp += compute_group_idx * compute_group_size;
+    }
+
+    n -= compute_group_idx * compute_group_size;
+
+    for (int i_start = 0; i_start < n && i_start < compute_group_size;
+         i_start += blockDim.x * PARALLEL_SIZE) {
+      MT r_g[PARALLEL_SIZE];
```

Comment:
Try to use `phi::AlignedVector` instead of the raw array. Keep the coding styles of the other codes in PaddlePaddle.